api/v1alpha2/awsmachine_conversion.go

```diff
@@ -19,23 +19,71 @@ package v1alpha2
 import (
 	apiconversion "k8s.io/apimachinery/pkg/conversion"
 	infrav1alpha3 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
 	"sigs.k8s.io/controller-runtime/pkg/conversion"
 )
 
 // ConvertTo converts this AWSMachine to the Hub version (v1alpha3).
 func (src *AWSMachine) ConvertTo(dstRaw conversion.Hub) error { // nolint
 	dst := dstRaw.(*infrav1alpha3.AWSMachine)
+
 	if err := Convert_v1alpha2_AWSMachine_To_v1alpha3_AWSMachine(src, dst, nil); err != nil {
 		return err
 	}
 
+	// Manually restore data from annotations
+	restored := &infrav1alpha3.AWSMachine{}
+	if ok, err := utilconversion.UnmarshalData(src, restored); err != nil || !ok {
+		return err
+	}
+
+	return restoreAWSMachineSpec(&restored.Spec, &dst.Spec)
+}
+
+func restoreAWSMachineSpec(restored *infrav1alpha3.AWSMachineSpec, dst *infrav1alpha3.AWSMachineSpec) error {
+
+	dst.ImageLookupBaseOS = restored.ImageLookupBaseOS
+
+	// restore CloudInit if it was on the annotation
+	if restored.CloudInit == nil {
+		return nil
+	}
+
+	if dst.CloudInit == nil {
+		dst.CloudInit = &infrav1alpha3.CloudInit{
+			Encrypted: restored.CloudInit.Encrypted,
+			SecretARN: restored.CloudInit.SecretARN,
+		}
+		return nil
+	}
+
+	// overwrite secretARN with hydrated version only if it doesn't exist
+	if dst.CloudInit.SecretARN == nil {
+		dst.CloudInit.SecretARN = restored.CloudInit.SecretARN
+	}
```

Comment:
I think you can simplify this to `dst.CloudInit = restored.CloudInit`. It will handle nil/not-nil correctly. And because this is a conversion from a2 to a3, the restored data is the only data we have.

Reply:
> And because this is a conversion from a2 to a3, the restored data is the only data we have.
Not if the fields are backported. In any case, have simplified and added comments to explain.